src/eos-util.c

```diff
@@ -0,0 +1,423 @@
+                         GCancellable *cancellable,
+                         GError **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GFile) child = NULL;
+
+  if (!ensure_is_ancestor (dir, file, error))
+    return FALSE;
+
+  if (!g_file_delete (file, cancellable, &local_error))
+    {
+      if (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        g_clear_error (&local_error);
+      else
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+    }
+
+  child = g_object_ref (file);
+  for (;;)
+    {
+      g_autoptr(GFile) parent = g_file_get_parent (child);
+
+      if (g_file_equal (dir, parent))
+        break;
+
+      if (!g_file_delete (parent, cancellable, &local_error))
+        {
+          if (!(g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND) ||
+                g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_EMPTY)))
+            {
+              g_propagate_error (error, g_steal_pointer (&local_error));
+              return FALSE;
+            }
+
+          break;
+        }
+
+      g_set_object (&child, parent);
+    }
+
+  return TRUE;
+}
+
+static gboolean
+create_directories_and_file (GFile *target,
+                             GBytes *contents,
+                             GCancellable *cancellable,
+                             GError **error)
+{
+  g_autoptr(GFile) target_parent = g_file_get_parent (target);
+  g_autoptr(GError) local_error = NULL;
+  gconstpointer raw;
+  gsize len;
+
+  if (!g_file_make_directory_with_parents (target_parent, cancellable, &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+      g_clear_error (&local_error);
+    }
+
+  raw = g_bytes_get_data (contents, &len);
+  return g_file_replace_contents (target,
+                                  raw,
+                                  len,
+                                  NULL,
+                                  FALSE,
+                                  G_FILE_CREATE_NONE,
+                                  NULL,
+                                  cancellable,
+                                  error);
+}
+
+gboolean
+eos_updater_save_or_delete  (GBytes *contents,
+                             GFile *dir,
+                             const gchar *filename,
+                             GCancellable *cancellable,
+                             GError **error)
+{
+  g_autoptr(GFile) target = g_file_get_child (dir, filename);
+  g_autoptr(GFile) target_parent = g_file_get_parent (target);
+
+  if (contents == NULL)
+    return delete_file_and_subdirs (dir, target, cancellable, error);
+
+  return create_directories_and_file (target, contents, cancellable, error);
+}
+
+static GFile *
+get_eos_extensions_path (OstreeRepo *repo)
```

Comment:
This function should probably be named something which indicates that we're getting a `GFile` instance as opposed to just a path.